server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -203,7 +208,8 @@ public DataStream(
         boolean failureStore,
         List<Index> failureIndices,
         boolean rolloverOnWrite,
-        @Nullable DataStreamAutoShardingEvent autoShardingEvent
+        @Nullable DataStreamAutoShardingEvent autoShardingEvent,
+        long failureStoreGeneration
```

Comment:
Food for thought, should we consider making this `Long` so it can be nullable when failure store is disabled? 

I am thinking that now we have a failure store generation `1` even when the failure store is disabled, or if in the future we lazily create the first failure store index. In these two cases, conceptually, I feel `null` might be better representing that we have no indices to back the failure store so no generation. 

Any thoughts?

Reply:
Hmm that's an interesting point. I personally don't foresee us using `failureStoreGeneration` to determine whether we have failure store indices or not. And then I'm not sure I'd consider the null-checking that we'd have to do in some places worth it. I have to say it's a tough one for me to take a side on this (possibly because the impact is relatively small here).

Reply:
I've given this some thought today, hence a slow response. There will indeed be times where we'll have data streams without failure stores enabled, and in those cases maybe it makes sense to not have a generation on it, but then that means:

* When we add the ability to enable failure stores on a data stream, we'll need to initialize the generation.
* Alternatively, if we lazily create the failure store, we'll need to remember to do it there.

These aren't too bad on their own, and I feel either route of leaving it set to 1 or nulling it out is fine. I think the question gets more interesting when you think about turning the feature on and off for a data stream which is another thing we're adding soon to support the feature. We'll need to support disabling failure stores for a data stream before the feature goes to GA. If you disable the feature for a data stream, does the stream hang on to the failure stores but not use them? Does it expel them from the data stream? In either case does it make sense to null the field out? If we hold on to the backing indices, probably not. If we expel them from the data stream, then if we turn the feature back on, what should the generation be? I'd assume it would be unchanged from when the feature was disabled. So it's possible to have the feature disabled on a data stream but still have a generation value on it if it's higher than the first generation.

I think overall, having it be an unboxed value is simpler in the long run (pun not intended), even if the feature is not on. Maybe it looks a little weird in the JSON representation, but we could always add some logic there to conditionally leave it off if the feature is disabled and the generation value of 1 is otherwise implied, then in the parser just set it to 1 if it's not present.

Reply:
+1 from me too